File: elixir/lib/wildling/token.ex
defmodule Wildling.Token do
  @moduledoc false

  defstruct src: "", start_length: 1, end_length: 1, variants: [], count: 0

  def new(options) when is_map(options) do
    src = Map.get(options, "src", Map.get(options, :src, ""))
    start_length = default_integer(Map.get(options, "startLength", Map.get(options, :startLength)), 1)
    end_length = default_integer(Map.get(options, "endLength", Map.get(options, :endLength)), 1)
    variants = Map.get(options, "variants", Map.get(options, :variants, [])) || []

    variant_count = length(variants)

    count =
      Enum.reduce(start_length..end_length, 0, fn len, acc ->
        acc + int_pow(variant_count, len)
      end)

    %__MODULE__{
      src: src,
      start_length: start_length,
      end_length: end_length,
      variants: variants,
      count: count
    }
  end

  def count(%__MODULE__{count: count}), do: count

  def src(%__MODULE__{src: src}), do: src

  def get(%__MODULE__{} = token, index) do
    cond do
      index > token.count - 1 or index < 0 ->
        ""

      index == 0 and token.start_length == 0 ->
        ""

      true ->
        variant_count = length(token.variants)

        {string_length, index_with_offset} =
          Enum.reduce_while(token.start_length..token.end_length, {token.start_length, index}, fn len,
                                                                                                 {_sl, idx} ->
            offset_count = int_pow(variant_count, len)

            if idx < offset_count do
              {:halt, {len, idx}}
            else
              {:cont, {len, idx - offset_count}}
            end
          end)

        build_string(token.variants, string_length, index_with_offset)
    end
  end

  defp build_string(variants, string_length, index) do
    variant_count = length(variants)

    {chars, _} =
      Enum.reduce(1..string_length, {[], index}, fn _, {acc, idx} ->
        variant_index = rem(idx, variant_count)
        {[Enum.at(variants, variant_index) | acc], div(idx, variant_count)}
      end)

    chars |> Enum.reverse() |> Enum.join()
  end

  defp default_integer(option, _fallback) when is_integer(option) and option >= 0, do: option
  defp default_integer(_, fallback), do: fallback

  defp int_pow(_base, 0), do: 1

  defp int_pow(base, exp) when is_integer(exp) and exp > 0 do
    Enum.reduce(1..exp, 1, fn _, acc -> acc * base end)
  end
end

File: lua/lib/wildling/token.lua
local M = {}

local function default_integer(option, fallback)
    if type(option) == "number" and option >= 0 and option == math.floor(option) then
        return option
    end
    return fallback
end

local function int_pow(base, exp)
    local result = 1
    for _ = 1, exp do
        result = result * base
    end
    return result
end

local Token = {}
Token.__index = Token

function Token.new(options)
    options = options or {}
    local self = setmetatable({}, Token)
    self._src = options.src or ""
    self._start_length = default_integer(options.startLength, 1)
    self._end_length = default_integer(options.endLength, 1)
    self._variants = options.variants or {}
    self._count = 0
    for length = self._start_length, self._end_length do
        self._count = self._count + int_pow(#self._variants, length)
    end
    return self
end

function Token:count()
    return self._count
end

function Token:src()
    return self._src
end

function Token:get(index)
    if index > self._count - 1 or index < 0 then
        return ""
    end

    if index == 0 and self._start_length == 0 then
        return ""
    end

    local index_with_offset = index
    local string_length = self._start_length
    for length = self._start_length, self._end_length do
        string_length = length
        local offset_count = int_pow(#self._variants, length)
        if index_with_offset < offset_count then
            break
        end
        index_with_offset = index_with_offset - offset_count
    end

    local string_array = {}
    for _ = 1, string_length do
        if #self._variants == 0 then
            break
        end
        local variant_index = index_with_offset % #self._variants
        index_with_offset = index_with_offset // #self._variants
        string_array[#string_array + 1] = self._variants[variant_index + 1]
    end
    return table.concat(string_array)
end

function M.create_token(options)
    return Token.new(options)
end

return M
